### PR TITLE
[IMP] mail: allow non-template editors users to use simple expressions

### PR DESCRIPTION
**Pull request by std-odoo**
Purpose
=======
Allow the non-template editor users to write simple expression
(e.g. object.name...) in mass-mailing but also in mail templates.

The mail composer in mass-mode is still in readonly mode if we are not
template editor, and we selected a template, because it might contain
complex code (conditions, loops, etc).

Task-3877116

### CODE REVIEW

**Comment by std-odoo** [comment]
@odony Hi :)

I did the regex rendering we talk about on discord, to not call the qweb rendering function (and safe eval for inline qweb). The code is not that complex in the end, so I changed my mind and I think it's a good idea :)

**Review by tbe-odoo — APPROVED** [review]
Hello, I have been testing this PR locally this morning and tried to break its security.
Haven't succeeded fortunately (or unfortunately??), regarding the code I have not much to say.

During my tests I figured out that for example if you use 
```xml
<t t-out="object.name"/>
```
It works well, but as soon as you add spaces at the tag closing like this:
```xml
<t t-out"object.name" / >
```
It is considered as forbidden. You might modify this to be less strict since I am pretty sure a lot of people will not understand why their modification doesn't work since they are not doing anything dangerous.

I approve from my end, but I think it need a second pair of eyes to see if I didn't catch something fishy.


Nice work and keep it up!!

**Comment by std-odoo** [comment]
@tbe-odoo Hi :) 

Thanks for testing :) 

I don't think we should allow
```
<t t-out"object.name" / >
```

Because it will crash in etree anyway
image

File: addons/web/static/src/views/fields/char/char_field.js [hunk 1]
```diff
@@ -53,13 +57,50 @@ export class CharField extends Component {
...
+    }
+
+    async onDynamicPlaceholderValidate(chain, defaultValue) {
+        if (chain) {
+            this.input.el.focus();
+            defaultValue = defaultValue.replace("|||", "");
+            let dynamicPlaceholder =
+                "object." +
+                chain +
+                (defaultValue && defaultValue !== "" ? ` ||| ${defaultValue}` : "");
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/char/char_field.js:88]: You could shorten it.
```suggestion
            const dynamicPlaceholder = `{{object.${chain}${
                defaultValue?.length ? ` ||| ${defaultValue}` : ""
            }}}`;
```

File: addons/web/static/src/views/fields/char/char_field.js [hunk 2]
```diff
@@ -53,13 +57,50 @@ export class CharField extends Component {
...
+                dynamicPlaceholder +
+                this.input.el.value.slice(this.selectionStart);
+
+            this.input.el.value = newValue;
+            // trigger events to make the field dirty
+            this.input.el.dispatchEvent(new InputEvent("input"));
+            this.input.el.dispatchEvent(new KeyboardEvent("keydown"));
+            this.input.el.focus();
+            const newCursorPos = this.selectionStart + dynamicPlaceholder.length;
+            this.input.el.setSelectionRange(newCursorPos, newCursorPos);
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/char/char_field.js:101]: Could you try:
```suggestion
            this.input.el.setRangeText(dynamicPlaceholder, this.selectionStart, this.selectionStart, "end");
```
To see if you have the same behavior? [Source](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setRangeText)

File: addons/web/static/src/views/fields/char/char_field.xml [hunk 3]
```diff
@@ -8,20 +8,30 @@
         <t t-else="">
             <input
                 class="o_input"
-                t-att-class="{'o_field_translate': isTranslatable}"
+                t-att-class="{'o_field_translate': isTranslatable,
+                    'o_field_placeholder': hasDynamicPlaceholder}"
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/char/char_field.xml:12]: Easier to read
```suggestion
                t-att-class="{
                    'o_field_translate': isTranslatable,
                    'o_field_placeholder': hasDynamicPlaceholder
                }"
```

File: addons/web/static/src/views/fields/dynamic_placeholder_hook.js [hunk 4]
```diff
@@ -27,8 +27,9 @@ export function useDynamicPlaceholder(elementRef) {
         // When the user cancel/close the popover, the path is empty.
         if (path) {
             let dynamicPlaceholder = "{{object." + path;
+            defaultValue = defaultValue.replace("|||", "");
             dynamicPlaceholder +=
-                defaultValue && defaultValue !== "" ? ` or '''${defaultValue}'''}}` : "}}";
+                defaultValue && defaultValue !== "" ? ` ||| ${defaultValue}}}` : "}}";
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/dynamic_placeholder_hook.js:32]:
```suggestion
            defaultValue = defaultValue.replace("|||", "");
            const dynamicPlaceholder = `{{object.${path}${
                defaultValue?.length ? ` ||| ${defaultValue}` : ""
            }}}`;
```

File: addons/web/static/src/views/fields/dynamic_placeholder_popover.js [hunk 5]
```diff
@@ -16,16 +17,36 @@ export class DynamicPlaceholderPopover extends Component {
...
         });
+
+        onWillStart(async () => {
+            this.isTemplateEditor = await user.hasGroup("mail.group_mail_template_editor");
+        });
     }
 
-    filter(fieldDef) {
+    filter(fieldDef, path) {
+        const fullPath = "object" + (path ? `.${path}` : "") + `.${fieldDef.name}`;
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/dynamic_placeholder_popover.js:27]:
```suggestion
        const fullPath = `object${path ? `.${path}` : ""}.${fieldDef.name}`;
```

File: addons/web/static/src/views/fields/dynamic_placeholder_popover.js [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import { useAutofocus } from "@web/core/utils/hooks";
 import { ModelFieldSelectorPopover } from "@web/core/model_field_selector/model_field_selector_popover";
-import { Component, useState } from "@odoo/owl";
+import { Component, useState, onWillStart } from "@odoo/owl";
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/dynamic_placeholder_popover.js:3]: :socks:

File: addons/web/static/src/views/fields/dynamic_placeholder_popover.xml [hunk 7]
```diff
@@ -4,41 +4,41 @@
...
+                        <span>or</span>
+                    </t>
                     <div class="o_model_field_selector_default_value_input mt-2">
                         <input type="text"
-                               placeholder='Type a default text or press ENTER'
-                               class="o_input"
-                               t-att-value="state.defaultValue"
-                               t-on-input="(ev) => this.setDefaultValue(ev.target.value)"
-                               t-ref="autofocus"/>
+                            placeholder="What shoud we write when the field is empty"
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/dynamic_placeholder_popover.xml:24]:
```suggestion
                            placeholder="What should we write when the field is empty"
```

File: addons/web/static/src/views/fields/dynamic_placeholder_popover.xml [hunk 8]
```diff
@@ -4,41 +4,41 @@
...
-                        </li>
-                    </ul>
+                <div class="o_model_field_selector_popover_footer border-top py-1 my-2 px-2">
+                    <input type="text" class="o_input o_model_field_selector_debug"
+                        disabled="disabled" t-att-value="state.path" />
                 </div>
-                <div class="o_model_field_selector_popover_footer border-top py-1 px-2">
-                    <input type="text" class="o_input o_model_field_selector_debug" disabled="disabled" t-att-value="state.path" />
+                <div class="d-flex flex-row justify-content-around py-2">
+                    <button class="btn btn-primary" t-on-click="() => this.validate()">
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/dynamic_placeholder_popover.xml:36]: I never know if you need the `.bind` or if it's not allowed on `t-on-` directives
```suggestion
                    <button class="btn btn-primary" t-on-click="validate">
```

File: addons/web_editor/static/src/js/backend/html_field.js
Comment (thjo-odoo) [inline addons/web_editor/static/src/js/backend/html_field.js:1]: Since the new editor is on the way you may need to also adapt the command in `html_editor`.

File: odoo/tools/rendering_tools.py [hunk 9]
```diff
@@ -42,33 +42,34 @@ def parse_inline_template(text):
...
     return groups
 
 def convert_inline_template_to_qweb(template):
     template_instructions = parse_inline_template(template or '')
     preview_markup = []
-    for string, expression in template_instructions:
+    for string, expression, default in template_instructions:
         if expression:
-            preview_markup.append(Markup('{}<t t-out="{}"/>').format(string, expression))
+            preview_markup.append(Markup('{}<t t-out="{}">{}</t>').format(string, expression, default))
```
Comment (thjo-odoo) [inline odoo/tools/rendering_tools.py:61]: Question:
Do we absolutely need to use the format on Markup, or could we use f-strings here?

**Review by thjo-odoo — COMMENTED** [review]
@std-odoo 

Hello, Gotta say it's quite the change around here. Nice work!
Here's the validation, if something seems unclear feel free to ask.

It also seems that your branch is in conflict, Good Luck.

Cheers!

**Review by std-odoo — COMMENTED** [review]
@thjo-odoo Hi :)

Thanks for the validation, I've done the changes and answered :)

Cheers

File: addons/web/static/src/views/fields/dynamic_placeholder_popover.js [hunk 10]
```diff
@@ -28,8 +48,9 @@ export class DynamicPlaceholderPopover extends Component {
         }
         this.props.close();
     }
-    setPath(path) {
+    setPath(path, fieldInfo) {
         this.state.path = path;
+        this.state.fieldName = fieldInfo && fieldInfo.string;
```
Comment (thjo-odoo) [inline addons/web/static/src/views/fields/dynamic_placeholder_popover.js:53]: You can shorten this to
```suggestion
        this.state.fieldName = fieldInfo?.string;
```
[Optional Chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) lets you do the same. 
Unless you absolutely need the value to be false and a *nullish* value doesn't work in your flow.

**Review by thjo-odoo — COMMENTED** [review]
@std-odoo 
Hello, here's the 2nd pass. From what I am reading there's no big problems left.
There's some details and alternative implementations.
Let me know if you have questions.
Cheers!

File: addons/mail/wizard/mail_compose_message_views.xml [hunk 11]
```diff
@@ -101,7 +105,10 @@
             <field name="arch" type="xml">
                 <form string="Templates">
                     <group>
-                        <field name="subject" placeholder="e.g. &quot;Welcome to MyCompany&quot; or &quot;Nice to meet you, {{ object.name }}&quot;"/>
+                        <field name="subject"
+                            options="{'dynamic_placeholder': true, 'dynamic_placeholder_model_reference_field': 'model'}"
+                            placeholder="e.g. &quot;Welcome to MyCompany&quot; or &quot;Nice to meet you, {{ object.name }}&quot;"/>
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/mail/wizard/mail_compose_message_views.xml:110]: I think it's correct like that, how should I change it ?

File: addons/web/static/src/views/fields/char/char_field.js [hunk 12]
```diff
@@ -53,13 +57,45 @@ export class CharField extends Component {
...
+            }}}`;
+            this.input.el.setRangeText(
+                dynamicPlaceholder,
+                this.selectionStart,
+                this.selectionStart,
+                "end"
+            );
+            // trigger events to make the field dirty
+            this.input.el.dispatchEvent(new InputEvent("input"));
+            this.input.el.dispatchEvent(new KeyboardEvent("keydown"));
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/web/static/src/views/fields/char/char_field.js:95]: no, you can try in the interface

**Review by std-odoo — COMMENTED** [review]
@thjo-odoo Hi :)

Changes done / answered :)

Cheers

File: addons/mail/models/mail_render_mixin.py [hunk 13]
```diff
@@ -332,6 +337,40 @@ def _render_template_qweb(self, template_src, model, res_ids,
 
         return results
 
+    @api.model
+    def _render_template_qweb_static(self, template_src, model, res_ids):
+        """Render the template with regex instead of qweb to avoid `eval` call."""
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_render_mixin.py:344]: Maybe worth a comment telling 'supporting only qweb allowed expression, no custom variables in that mode'

File: addons/mail/models/mail_render_mixin.py [hunk 14]
```diff
@@ -302,6 +303,10 @@ def _render_template_qweb(self, template_src, model, res_ids,
         if add_context:
             variables.update(**add_context)
 
+        if not self._has_complex_expression_template_qweb(template_src):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_render_mixin.py]: Can be called before variables computation, as they are not used in this mode

File: addons/mail/models/mail_render_mixin.py [hunk 15]
```diff
@@ -332,6 +337,40 @@ def _render_template_qweb(self, template_src, model, res_ids,
 
         return results
 
+    @api.model
+    def _render_template_qweb_static(self, template_src, model, res_ids):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_render_mixin.py:343]: Weird to call "static" something that is dynamic ^^ just a safer version of the other one.

maybe ``_render_template_qweb_regex`` (as I have no clear word for 'light', 'simple' in mind, maybe the implementation choice can be used as suffix).

or maybe ``_render_template_qweb_simple``

File: addons/mail/models/mail_render_mixin.py [hunk 16]
```diff
@@ -332,6 +337,40 @@ def _render_template_qweb(self, template_src, model, res_ids,
...
+        """Render the template with regex instead of qweb to avoid `eval` call."""
+        records = self.env[model].browse(res_ids)
+        result = {}
+        for record in records:
+            def replace(match):
+                tag = match.group(1)
+                expr = match.group(3)
+                default = match.group(9)
+                if not self.env['ir.qweb']._is_expression_allowed(expr):
+                    raise SyntaxError("Invalid expression for the static mode")
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_render_mixin.py:353]: Maybe include expression in error ? Always handy to know what is wrong ^^

File: addons/mail/models/mail_render_mixin.py [hunk 17]
```diff
@@ -332,6 +337,40 @@ def _render_template_qweb(self, template_src, model, res_ids,
...
+
+                value = escape(value or '')
+                return value if tag.lower() == 't' else f"<{tag}>{value}</{tag}>"
+
+            # normalize the HTML (add a parent div to avoid modification of the template
+            # it will be removed by html_normalize)
+            template_src = html_normalize(f'<div>{template_src}</div>')
+
+            result[record.id] = Markup(re.sub(
+                r'''<(\w+)[\s|\n]+t-out=[\s|\n]*(\'|\")((\w|\.)+)(\2)[\s|\n]*((\/>)|(>[\s|\n]*([^<>]*?))[\s|\n]*<\/\1>)''',
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_render_mixin.py:368]: I trust you (and the previous reviewers) on this one.

File: addons/mail/models/mail_render_mixin.py [hunk 18]
```diff
@@ -442,6 +484,27 @@ def _render_template_inline_template(self, template_txt, model, res_ids,
 
         return results
 
+    @api.model
+    def _render_template_inline_template_static(self, template_txt, model, res_ids):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_render_mixin.py:490]: Same proposal for "suffix" of method name

File: addons/mail/models/mail_render_mixin.py [hunk 19]
```diff
@@ -442,6 +484,27 @@ def _render_template_inline_template(self, template_txt, model, res_ids,
...
+        template = parse_inline_template(str(template_txt))
+        records = self.env[model].browse(res_ids)
+        result = {}
+        for record in records:
+            renderer = []
+            for string, expression, default in template:
+                renderer.append(string)
+                if expression:
+                    if not self.env['ir.qweb']._is_expression_allowed(expression):
+                        raise SyntaxError("Invalid expression for the static mode")
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_render_mixin.py:501]: Would include invalid expression in error

File: addons/mail/models/mail_render_mixin.py [hunk 20]
```diff
@@ -442,6 +484,27 @@ def _render_template_inline_template(self, template_txt, model, res_ids,
...
+                    try:
+                        value = reduce(lambda rec, field: rec[field], expression.split('.')[1:], record) or default
+                    except KeyError:
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_render_mixin.py:504]: Note that other rendering methods are smoother with error management (aka raising an user error in all cases with template source, ...). Is it intended to be "hasher" in those simpler versions ?

File: addons/mail/tests/test_mail_render.py [hunk 21]
```diff
@@ -409,6 +409,96 @@ def test_replace_local_links(self):
             rendered = self.env['mail.render.mixin']._replace_local_links(source)
             self.assertEqual(rendered, expected)
 
+    def test_qweb_static_rendering(self):
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_render.py:412]: :socks: same here, if we change "static" by a better suffix, don't forget to propagate :)

Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_render.py:412]: Could make a specific test class TestSimplifiedRendering (with a test tag wisely chosen) so that people are able to run tests for this rendering without having to run all other tests maybe ? I quite like having a bit more classes allowing to split a bit big files :)

File: addons/mail/tests/test_mail_render.py [hunk 22]
```diff
@@ -409,6 +409,96 @@ def test_replace_local_links(self):
...
+        o_qweb_render = self.env['ir.qweb']._render
+        for template, expected in static_templates:
+            with (patch('odoo.addons.base.models.ir_qweb.IrQWeb._render', side_effect=o_qweb_render) as qweb_render,
+                patch('odoo.addons.base.models.ir_qweb.unsafe_eval', side_effect=eval) as unsafe_eval):
+                self.assertEqual(render(template), expected)
+                self.assertFalse(qweb_render.called)
+                self.assertFalse(unsafe_eval.called)
+
+        # double check that we are able to catch the eval
+        non_static_templates = (
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_render.py:456]: Could include a test for context usage, so that we know currently it is considered as unsafe

File: addons/mail/tests/test_mail_render.py [hunk 23]
```diff
@@ -409,6 +409,96 @@ def test_replace_local_links(self):
...
+            ('''{{object.name ||| Default}}''', 'Alice'),
+            ('''{{object.contact_name ||| Default}}''', 'Default'),
+        )
+        for template, expected in static_templates:
+            with patch('odoo.tools.safe_eval.unsafe_eval', side_effect=eval) as unsafe_eval:
+                self.assertEqual(render(template), expected)
+                self.assertFalse(unsafe_eval.called)
+                self.assertFalse(self.env['mail.render.mixin']._has_complex_expression_template_inline_template(template))
+
+        non_static_templates = (
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_render.py:492]: Could include a test for context usage, so that we know currently it is considered as unsafe

File: addons/mail/tests/test_mail_template.py [hunk 24]
```diff
@@ -77,19 +77,26 @@ def test_mail_template_acl(self):
         self.assertFalse(self.user_employee.has_group('mail.group_mail_template_editor'))
         self.assertFalse(self.user_employee.has_group('base.group_sanitize_override'))
 
+        record = self.user_employee.partner_id
+
         # Group System can create / write / unlink mail template
-        mail_template = self.env['mail.template'].with_user(self.user_admin).create({'name': 'Test template'})
+        mail_template = self.env['mail.template'].with_user(self.user_admin).create({
+            'name': 'Test template',
+            'model_id': self.env['ir.model']._get('res.partner').id,
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_template.py]: :socks: ``_get_id`` :)

File: addons/mail/tests/test_mail_template.py [hunk 25]
```diff
@@ -77,19 +77,26 @@ def test_mail_template_acl(self):
         self.assertFalse(self.user_employee.has_group('mail.group_mail_template_editor'))
         self.assertFalse(self.user_employee.has_group('base.group_sanitize_override'))
 
+        record = self.user_employee.partner_id
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_template.py:81]: test_mail_template_acl would probably benefit from being split into several sub tests, but hey that can be done "later" (not really something necessary for the freeze :D )

**Review by tde-banana-odoo — APPROVED** [review]
Thanks for your work, it is effectively a nice help for people writing simple dynamic templates :) . LGTM, added some lint / naming remarks while going through, did not find much important to add :) .

Thanks !

**Comment by tbe-odoo** [comment]
@robodoo override=ci/security

**Review by std-odoo — COMMENTED** [review]
@tde-banana-odoo Hi :) Changes done :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
